crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -152,9 +152,22 @@ where
     }
 
     fn header_td_by_number(&self, number: BlockNumber) -> ProviderResult<Option<U256>> {
-        self.database.header_td_by_number(number)
+        if let Some(td) = self.database.header_td_by_number(number)? {
+            Ok(Some(td))
+        } else {
+            let block_hash = self.canonical_in_memory_state.state_by_number(number)
+                .map(|state| state.block().block().hash());
+            if let Some(hash) = block_hash {
+                self.database.header_td(&hash)
+            } else {
```

Comment:
```suggestion
            if let Some(hash) = self.canonical_in_memory_state.state_by_number(number)
                .map(|state| state.block().block().hash()) {
                self.database.header_td(&hash)
            } else {
```

collapse this

Reply:
@emhane done, can you please check it now